server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportDeleteDesiredBalanceAction.java

```diff
@@ -0,0 +1,127 @@
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import static org.elasticsearch.cluster.routing.allocation.allocator.AllocationActionListener.rerouteCompletionIsNotRequired;
+
+public class TransportDeleteDesiredBalanceAction extends TransportMasterNodeAction<DesiredBalanceRequest, ActionResponse.Empty> {
+
+    @Nullable
+    private final MasterServiceTaskQueue<ResetDesiredBalanceTask> resetDesiredBalanceTaskQueue;
+
+    @Inject
+    public TransportDeleteDesiredBalanceAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        AllocationService allocationService,
+        ShardsAllocator shardsAllocator
+    ) {
+        super(
+            DeleteDesiredBalanceAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            DesiredBalanceRequest::new,
+            indexNameExpressionResolver,
+            in -> ActionResponse.Empty.INSTANCE,
+            ThreadPool.Names.MANAGEMENT
+        );
+
+        this.resetDesiredBalanceTaskQueue = shardsAllocator instanceof DesiredBalanceShardsAllocator allocator
+            ? clusterService.createTaskQueue(
+                "reset-desired-balance",
+                Priority.NORMAL,
+                new ResetDesiredBalanceClusterExecutor(allocationService, allocator)
+            )
+            : null;
+    }
+
+    public record ResetDesiredBalanceTask(ActionListener<Void> listener) implements ClusterStateTaskListener {
+
+        @Override
+        public void onFailure(Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    private static final class ResetDesiredBalanceClusterExecutor implements ClusterStateTaskExecutor<ResetDesiredBalanceTask> {
+
+        private final AllocationService allocationService;
+        private final DesiredBalanceShardsAllocator desiredBalanceShardsAllocator;
+
+        ResetDesiredBalanceClusterExecutor(
+            AllocationService allocationService,
+            DesiredBalanceShardsAllocator desiredBalanceShardsAllocator
+        ) {
+            this.allocationService = allocationService;
+            this.desiredBalanceShardsAllocator = desiredBalanceShardsAllocator;
+        }
+
+        @Override
+        public ClusterState execute(BatchExecutionContext<ResetDesiredBalanceTask> batchExecutionContext) {
+            var state = batchExecutionContext.initialState();
+            desiredBalanceShardsAllocator.resetDesiredBalance();
+            state = allocationService.reroute(state, "reset-desired-balance", rerouteCompletionIsNotRequired());
+            for (var taskContext : batchExecutionContext.taskContexts()) {
+                taskContext.success(() -> taskContext.getTask().listener.onResponse(null));
```

Comment:
I think it would be helpful to delay returning from this API until we've computed the new balance and started the reconciliation, which I think would need to enqueue a listener in the allocator.